File: app/src/main/java/com/jakubwyc/mysecretvault/presenter/CreateVaultPresenter.java
package com.jakubwyc.mysecretvault.presenter;

import android.text.TextUtils;

import com.jakubwyc.mysecretvault.R;
import com.jakubwyc.mysecretvault.model.User;
import com.jakubwyc.mysecretvault.repository.Repository;
import com.jakubwyc.mysecretvault.view.CreateVaultView;

import rx.subscriptions.CompositeSubscription;

import static com.jakubwyc.mysecretvault.Utils.sha256;

public class CreateVaultPresenter implements Presenter<CreateVaultView> {

    private CreateVaultView view;
    private CompositeSubscription subscriptions = new CompositeSubscription();

    private String login;
    private String password;
    private String passwordRepeat;

    @Override
    public void attachView(CreateVaultView view) {
        this.view = view;
    }

    @Override
    public void detachView() {
        subscriptions.clear();
        view = null;
    }

    public void onLoginChanged(final CharSequence value) {
        login = value.toString();
    }

    public void onPasswordChanged(final CharSequence value) {
        password = value.toString();
    }

    public void onPasswordRepeatChanged(final CharSequence value) {
        passwordRepeat = value.toString();
    }

    public void createVault() {
        if (isUserDataValid()) {
            User user = new User(login, sha256(password));
            saveUser(user);
        }
    }

    private boolean isUserDataValid() {
        boolean valid = false;
        if (TextUtils.isEmpty(login)) {
            view.showToast(R.string.login_empty);
        } else if (TextUtils.isEmpty(password)) {
            view.showToast(R.string.password_empty);
        } else if (TextUtils.isEmpty(passwordRepeat)) {
            view.showToast(R.string.password_repeat_empty);
        } else if (!TextUtils.equals(password, passwordRepeat)) {
            view.showToast(R.string.passwords_not_equal);
        } else {
            valid = true;
        }
        return valid;
    }

    private void saveUser(User user) {
        subscriptions.add(Repository.getInstance().saveUser(user)
                .subscribe($ -> view.showToast(R.string.user_saved))
        );
    }
}

File: app/src/main/java/com/jakubwyc/mysecretvault/vault/create/CreateVaultPresenter.kt
package com.jakubwyc.mysecretvault.vault.create

import com.jakubwyc.mysecretvault.Presenter
import com.jakubwyc.mysecretvault.SystemContext
import com.jakubwyc.mysecretvault.model.User
import com.jakubwyc.mysecretvault.sha256
import rx.subscriptions.CompositeSubscription

class CreateVaultPresenter(val systemContext: SystemContext) : Presenter<CreateVaultView> {

    private var view: CreateVaultView? = null
    private val subscriptions = CompositeSubscription()

    private var login: String = ""
    private var password: String = ""
    private var passwordRepeat: String = ""
    private val userRepository
        get() = systemContext.userRepository

    override fun attachView(view: CreateVaultView) {
        this.view = view
    }

    override fun detachView() {
        subscriptions.clear()
        view = null
    }

    fun onLoginChanged(value: CharSequence) {
        login = value.toString()
    }

    fun onPasswordChanged(value: CharSequence) {
        password = value.toString()
    }

    fun onPasswordRepeatChanged(value: CharSequence) {
        passwordRepeat = value.toString()
    }

    fun createVault() {
        if (isUserDataValid) {
            val user = User(login, sha256(password))
            saveUser(user)
        }
    }

    private val isUserDataValid: Boolean
        get() {
            var valid = false
            when {
                login.isBlank() -> view?.showLoginEmptyToast()
                password.isBlank() -> view?.showPasswordEmptyToast()
                passwordRepeat.isBlank() -> view?.showPasswordRepeatEmptyToast()
                password != passwordRepeat -> view?.showPasswordsNotEqualToast()
                else -> valid = true
            }
            return valid
        }

    private fun saveUser(user: User) {
        subscriptions.add(userRepository.saveUser(user)
                .subscribe {
                    view?.showUserSavedToast()
                    view?.goToMainScreen()
                }
        )
    }
}
